Real nvp floating num masked

hartikainen opened the pull request:
As mentioned in #662, it would be nice if the number of masked elements in RealNVP could be defined as proportions instead of just integers.

Happy to take feedback on this.

hartikainen commented on tensorflow_probability/python/bijectors/real_nvp.py:
```diff
@@ -206,24 +209,37 @@ def _bijector_fn(x0, input_depth, **condition_kwargs):
+    if isinstance(self._num_masked, (int, np.integer)):
+      return self._num_masked
+    elif if np.issubdtype(self._num_masked, np.floating):
+      return np.floor(self._input_depth * self._num_masked)
```
Should this be `floor` for both negative and positive values?

hartikainen commented on tensorflow_probability/python/bijectors/real_nvp.py:
```diff
@@ -171,14 +176,12 @@ def __init__(self,
       name: Python `str`, name given to ops managed by this object.
 
     Raises:
-      ValueError: If num_masked < 0.
```
This is incorrect and should be removed even if this PR doens't get merged.

hartikainen commented on tensorflow_probability/python/bijectors/real_nvp.py:
```diff
@@ -206,24 +209,37 @@ def _bijector_fn(x0, input_depth, **condition_kwargs):
+    if isinstance(self._num_masked, (int, np.integer)):
+      return self._num_masked
+    elif np.issubdtype(type(self._num_masked), np.floating):
+      return int(np.floor(self._input_depth * self._num_masked))
```
Should this be floor for both negative and positive values?

SiegeLordEx replied:
I think we should round towards zero, so flipping the sign doesn't alter the number of units masked.

SiegeLordEx commented on tensorflow_probability/python/bijectors/real_nvp.py:
```diff
@@ -144,10 +145,14 @@ def __init__(self,
-        should be masked. Must be in the closed interval `[0, D-1]`, where `D`
-        is the event size of the base distribution. If the value is negative,
-        then the last `d` units of the event are masked instead.
+      num_masked: Python `int` or `float` indicating the number of units of the
```
I'm not super happy with reusing the argument for this purpose, can we add a second argument (e.g. `fraction_masked` and use that instead?).

SiegeLordEx commented on tensorflow_probability/python/bijectors/real_nvp.py:
```diff
@@ -144,10 +145,14 @@ def __init__(self,
+        event are masked instead. If `float`, must be in the closed interval
+        `[-1, 1]`, and the value represents the proportion of the values to be
+        masked. If negative, then the last proportion of units will of the
+        are masked instead.
```
This needs to say how we round it.

SiegeLordEx commented on tensorflow_probability/python/bijectors/real_nvp.py:
```diff
@@ -206,24 +209,37 @@ def _bijector_fn(x0, input_depth, **condition_kwargs):
         name=name)
 
+  @property
+  def num_masked(self):
```
I don't think we need for this to be public, would you mind making this private instead? I'm particularly concerned that this will raise an error if `_input_depth` is `None`. We could document this, but if we don't need to have this public, then there's no need to worry about this.

Could make it a `_get_num_masked()` to avoid the name collision.

hartikainen commented:
Hey @SiegeLordEx, thanks a lot for the feedback! I've addressed your comments and split the `num_masked` into two different arguments. Let me know if there's anything else that you think could be improved.

hartikainen commented on tensorflow_probability/python/bijectors/real_nvp.py:
```diff
@@ -171,20 +181,36 @@ def __init__(self,
+      self._fraction_masked = float(fraction_masked)
+      self._reverse_mask = self._fraction_masked < 0
+
+    if shift_and_log_scale_fn is not None and bijector_fn is not None:
```
Note that these changes are not strictly needed for this PR. But I think this way of checking the `None`s is better practice than what it used to be (`bool(shift_and_log_scale_fn) == bool(bijector_fn)`).

hartikainen commented on tensorflow_probability/python/bijectors/real_nvp_test.py:
```diff
@@ -60,25 +64,39 @@ def testBijectorWithTrivialTransform(self):
+      _, x2_ = np.split(x_, [input_depth - abs(flip_nvp._masked_size)], axis=-1)
+
+      # Check latter half is the same after passing thru reversed mask RealNVP.
+      _, forward_x2 = tf.split(
```
Not sure about the linting here. Is there a command I can run to lint with the tfp styles?

SiegeLordEx replied:
You can do `pylint --rcfile=testing/pylintrc <python files>`.

SiegeLordEx replied:
It's still not happy about this line. I think you'll probably need to disable the lint here as it seems like a false positive. Just add a '  # pylint: disable=<lint name' here.

hartikainen commented:
The linting should be fixed now!

SiegeLordEx commented:
Thanks! Looks like our Travis build is kind of borked, so I'll take it from here. I'll send it to our internal review system and once that is merged, this'll get closed automatically.

hartikainen commented:
Nice, thanks @SiegeLordEx!
